--- zygote/zygote_main_linux.cc	2016-02-17 11:54:07.573103888 +0000
+++ zygote/zygote_main_linux.cc	2016-02-17 12:10:14.451183368 +0000
@@ -40,6 +40,7 @@
 #include "content/public/common/sandbox_linux.h"
 #include "content/public/common/zygote_fork_delegate_linux.h"
 #include "content/zygote/zygote_linux.h"
+#include "crypto/nss_util.h"
 #include "sandbox/linux/services/credentials.h"
 #include "sandbox/linux/services/init_process_reaper.h"
 #include "sandbox/linux/services/namespace_sandbox.h"
@@ -332,9 +333,15 @@
   // cached and there's no more need to access the file system.
   scoped_ptr<icu::TimeZone> zone(icu::TimeZone::createDefault());
 
+#if defined(USE_NSS_CERTS)
+  // NSS libraries are loaded before sandbox is activated. This is to allow
+  // successful initialization of NSS which tries to load extra library files.
+  crypto::LoadNSSLibraries();
+#elif defined(USE_OPENSSL)
   // Pass BoringSSL a copy of the /dev/urandom file descriptor so RAND_bytes
   // will work inside the sandbox.
   RAND_set_urandom_fd(base::GetUrandomFD());
+#endif
 
 #if defined(ENABLE_PLUGINS)
   // Ensure access to the Pepper plugins before the sandbox is turned on.
