--- codec/png_codec.cc	2019-05-14 22:01:49.136124683 +0200
+++ codec/png_codec.cc	2019-05-14 22:12:32.236126619 +0200
@@ -9,7 +9,10 @@
 #include "base/logging.h"
 #include "base/macros.h"
 #include "base/strings/string_util.h"
+#if 0
 #include "third_party/libpng/png.h"
+#endif
+#include "libpng12/png.h"
 #include "third_party/skia/include/core/SkBitmap.h"
 #include "third_party/skia/include/core/SkColorPriv.h"
 #include "third_party/skia/include/core/SkUnPreMultiply.h"
