--- platform/window_manager_wayland.cc	2016-03-11 10:25:04.514719111 +0000
+++ platform/window_manager_wayland.cc	2016-03-11 10:25:13.966852014 +0000
@@ -14,7 +14,7 @@
 #include "ozone/platform/ozone_gpu_platform_support_host.h"
 #include "ozone/platform/ozone_wayland_window.h"
 #include "ozone/wayland/ozone_wayland_screen.h"
-#include "ui/aura/window.h"
+//#include "ui/aura/window.h"
 #include "ui/events/event_utils.h"
 #include "ui/events/ozone/layout/keyboard_layout_engine_manager.h"
 #include "ui/events/platform/platform_event_source.h"
