--- ozone/layout/xkb/xkb_keyboard_layout_engine.cc	2019-05-15 01:18:28.783160203 +0200
+++ ozone/layout/xkb/xkb_keyboard_layout_engine.cc	2019-05-15 01:19:46.867160438 +0200
@@ -693,7 +693,12 @@
       base::Bind(&LoadKeymap, layout_name, base::ThreadTaskRunnerHandle::Get(),
                  reply_callback));
 #else
-  NOTIMPLEMENTED();
+  xkb_keymap* keymap = xkb_map_new_from_string(
+      xkb_context_.get(), layout_name.c_str(), XKB_KEYMAP_FORMAT_TEXT_V1,
+      XKB_KEYMAP_COMPILE_NO_FLAGS);
+  if (!keymap)
+    return false;
+  SetKeymap(keymap);
 #endif  // defined(OS_CHROMEOS)
   return true;
 }
