--- browser/layout_test/layout_test_message_filter.cc	2016-02-26 18:03:06.791075885 +0000
+++ browser/layout_test/layout_test_message_filter.cc	2016-02-26 18:03:44.893633162 +0000
@@ -9,7 +9,7 @@
 #include "content/public/browser/child_process_security_policy.h"
 #include "content/public/browser/permission_type.h"
 #include "content/public/test/layouttest_support.h"
-#include "content/shell/browser/layout_test/layout_test_bluetooth_adapter_provider.h"
+//#include "content/shell/browser/layout_test/layout_test_bluetooth_adapter_provider.h"
 #include "content/shell/browser/layout_test/layout_test_browser_context.h"
 #include "content/shell/browser/layout_test/layout_test_content_browser_client.h"
 #include "content/shell/browser/layout_test/layout_test_notification_manager.h"
@@ -168,9 +168,11 @@
 
 void LayoutTestMessageFilter::OnSetBluetoothAdapter(const std::string& name) {
   DCHECK_CURRENTLY_ON(BrowserThread::UI);
+#if 0
   SetBluetoothAdapter(
       render_process_id_,
       LayoutTestBluetoothAdapterProvider::GetBluetoothAdapter(name));
+#endif
 }
 
 }  // namespace content
--- browser/shell_content_browser_client.cc	2016-02-26 17:57:18.439980642 +0000
+++ browser/shell_content_browser_client.cc	2016-02-26 17:58:09.158722531 +0000
@@ -19,7 +19,7 @@
 #include "content/public/common/content_switches.h"
 #include "content/public/common/url_constants.h"
 #include "content/public/common/web_preferences.h"
-#include "content/public/test/test_mojo_app.h"
+//#include "content/public/test/test_mojo_app.h"
 #include "content/shell/browser/blink_test_controller.h"
 #include "content/shell/browser/layout_test/layout_test_browser_main_parts.h"
 #include "content/shell/browser/layout_test/layout_test_resource_dispatcher_host_delegate.h"
@@ -217,8 +217,10 @@
 
 void ShellContentBrowserClient::RegisterOutOfProcessMojoApplications(
       OutOfProcessMojoApplicationMap* apps) {
+#if 0
   apps->insert(std::make_pair(GURL(kTestMojoAppUrl),
                               base::UTF8ToUTF16("Test Mojo App")));
+#endif
 }
 
 void ShellContentBrowserClient::AppendExtraCommandLineSwitches(
--- utility/shell_content_utility_client.cc	2016-02-26 18:14:19.306878134 +0000
+++ utility/shell_content_utility_client.cc	2016-02-26 18:14:53.986383172 +0000
@@ -6,7 +6,7 @@
 
 #include "base/bind.h"
 #include "base/memory/scoped_ptr.h"
-#include "content/public/test/test_mojo_app.h"
+//#include "content/public/test/test_mojo_app.h"
 #include "mojo/shell/static_application_loader.h"
 
 namespace content {
@@ -14,7 +14,7 @@
 namespace {
 
 scoped_ptr<mojo::ApplicationDelegate> CreateTestApp() {
-  return scoped_ptr<mojo::ApplicationDelegate>(new TestMojoApp);
+//  return scoped_ptr<mojo::ApplicationDelegate>(new TestMojoApp);
 }
 
 }  // namespace
@@ -24,8 +24,10 @@
 
 void ShellContentUtilityClient::RegisterMojoApplications(
     StaticMojoApplicationMap* apps) {
+#if 0
   apps->insert(
       std::make_pair(GURL(kTestMojoAppUrl), base::Bind(&CreateTestApp)));
+#endif
 }
 
 }  // namespace content
